packages/effection/src/context.js

```diff
@@ -110,52 +118,97 @@ Thanks!`);
   }
 
   resume(value) {
-    if (this.isRunning) {
-      this.result = value;
-    }
-    if (this.requiredChildren.size > 0) {
-      this.state = 'waiting';
-    } else {
-      this.finalize('completed', value);
+    if(this.isBlocking) {
+      if (this.isRunning) {
+        this.result = value;
+        if(contextOf(value)) {
+          this.link(contextOf(value));
+        }
+      }
+      if(Array.from(this.children).some((c) => (c === value) ? this.blockOnReturnedContext : c.isRequired)) {
+        this.state = 'waiting';
+      } else {
+        this.finalize('completed', value);
+      }
     }
   }
 
   fail(error) {
-    this.finalize('errored', error);
+    if(this.isBlocking) {
+      this.finalize('errored', error);
+    }
   }
 
   finalize(state, result) {
-    this.halt = this.resume = this.fail = this.finalize = function noop() {};
-
-    this.state = state;
-    this.result = result || this.result;
-
-    for (let child of [...this.children].reverse()) {
-      child.halt(result);
-    }
+    if(this.isBlocking) {
+      this.state = state;
+      this.result = result || this.result;
+
+      for (let child of [...this.children].reverse()) {
+        if(this.blockOnReturnedContext || contextOf(this.result) !== child) {
+          child.halt(result);
+        }
+      }
 
-    for (let hook of [...this.exitHooks].reverse()) {
-      try {
-        hook();
-      } catch(e) {
-        /* eslint-disable no-console */
-        console.error(`
+      for (let hook of [...this.exitHooks].reverse()) {
+        try {
+          hook();
+        } catch(e) {
+          /* eslint-disable no-console */
+          console.error(`
 CRITICAL ERROR: an exception was thrown in an exit handler, this might put
 Effection into an unknown state, and you should avoid this ever happening.
 Original error:`);
-        console.error(e);
-        /* eslint-enable no-console */
+          console.error(e);
+          /* eslint-enable no-console */
+        }
       }
+
+      if (this.parent) {
+        this.parent.trapExit(this);
+      }
+
+      this.finalizePromise();
     }
+  }
+
+  trapExit(child) {
+    this.unlink(child);
 
-    if (this.parent) {
-      this.parent.children.delete(this);
-      this.parent.requiredChildren.delete(this);
+    if(child.isCompleted && contextOf(child.result)) {
+      this.link(contextOf(child.result));
     }
 
-    this.finalizePromise();
+    if(child.isErrored) {
+      this.fail(child.result);
+    } else if (this.isWaiting && Array.from(this.children).every((c) => !c.isRequired)) {
+      this.finalize('completed');
+    }
+  }
+
+  link(child) {
+    if(this.id === child.id) {
+      throw new Error('cannot link context to itself');
+    }
```

Comment:
Is it worth adding in a check here to make sure that the new parent is higher in the tree than the old parent? It feels like it is not ok, to transfer ownership "down-tree", although we may want a way to do that explicitly in the future.

Reply:
That was sort of what I was thinking, I think it makes sense to expose `link` as an operation, to essentially "move" a child somewhere, though there might be some weird implications to that which I haven't thought through, so I didn't add it in this PR. But we could definitively add such a check for now, until we're sure that it is actually safe to do this.